### PR TITLE
[CPU]The shape of the memory descriptor is considered in selectPreferPrimitiveDescriptor of Subgraph

### PR DESCRIPTION
**Pull request by xipingyan**
### Details:
 - *selectPreferPrimitiveDescriptor take into account shapes of the memory descriptors, because scalar shape node's reorder has less computation*
 - *New logic only work for Subgraph*

### Tickets:
 - *137307*
 - *139904*

### CODE REVIEW

File: src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/subgraph_select_pd.cpp [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+*/
+
+class SubgraphSelectPD : virtual public SubgraphBaseStaticTest {
+protected:
+    void SetUp() override {
+        targetDevice = ov::test::utils::DEVICE_CPU;
+        abs_threshold = 1e-2;
+
+        auto type = element::f32;
+#define CONST1 128
```
Comment (praasz) [inline src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/subgraph_select_pd.cpp:47]: Can be `constexpr` variable instead `define`?

**Comment by dmitry-gorokhov** [comment]
@maxnick  @EgorDuplensky  Please take a look

**Comment by xipingyan** [comment]
@EgorDuplensky Please help review again. Thank you very much.

File: src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/subgraph_select_pd.cpp [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+               \                       /                            /
+               Subgraph(f32_acdb_{1,64,30,30})
+                            |
+                            |
+               Convolution(f32_acdb_{1,1,28,28})
+                            |
+                          Result
+    The Subgraph node have 3 inputs: they don't have same layout.
+    Expected: Reorder is inserted after VariadicSplit[0] and VariadicSplit[1], not inserted after MVN.
+    Because VariadicSplit's output layout is [1,64,1,1], it' reorder almost have no calculation.
```
Comment (EgorDuplensky) [inline src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/subgraph_select_pd.cpp:37]: Is it really about "almost no calculation" rather than less compute in comparison with other Reorder insertion options?

File: src/plugins/intel_cpu/src/node.cpp [hunk 3]
```diff
@@ -316,6 +315,103 @@ void Node::selectPreferPrimitiveDescriptor(const std::vector<impl_desc_type>& pr
...
+    for (auto s : pshape) {
+        if (s.is_static() && s.get_length() == 1) {
+            oneNum++;
+        }
+    }
+    return oneNum >= sz - 1;
+}
+
+void Node::selectPreferPrimitiveDescriptorWithShape(const std::vector<impl_desc_type>& priority, bool ignoreConstInputs) {
+    for (auto& type : priority) {
```
Comment (EgorDuplensky) [inline src/plugins/intel_cpu/src/node.cpp:330]: Could you please extract logical pieces of declarative actions from this gigantic multibranch triple nesting 'for loop' and define them as named functions / lambdas?
We better to have it as something like:
``` c++
void Node::selectPreferPrimitiveDescriptorWithShape(const std::vector<impl_desc_type>& priority, bool ignoreConstInputs) {
    for (auto& type : priority) {
        auto something = getSomethingFrom(thing);
        doSomethingWith(something);
        ...
    }
}
```
Currently the logic of this function is really hard to follow

**Review by EgorDuplensky — COMMENTED** [review]
Please update a commit message subject and PR name, so it contains relevant info regarding the actual change.
Please update a commit message body and PR description with the details of the change

File: src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/subgraph_select_pd.cpp (hunk 2 above)
Comment (xipingyan) [reply src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/subgraph_select_pd.cpp:37]: Function calls themselves take time, and sequential copying also takes time, so perhaps "less computation" is more accurate.

File: src/plugins/intel_cpu/src/node.cpp (hunk 3 above)
Comment (xipingyan) [reply src/plugins/intel_cpu/src/node.cpp]: Yes, it's hard to read, so I use lambda function replace loop.

File: src/plugins/intel_cpu/src/node.cpp [hunk 4]
```diff
@@ -316,6 +315,118 @@ void Node::selectPreferPrimitiveDescriptor(const std::vector<impl_desc_type>& pr
...
+                DEBUG_LOG(getName(), " pd[", i, "].inConfs[", j, "]"
+                          " is ", (isCompatible ? "compatible" : "not compatible"),
+                          " shape is ", (isScalarShape ? "scalar" : "not scalar"),
+                          " with parent ", parentPtr->getName(),
+                          " outConfs[", inNum, "], reorderLocalCostScore add to ", reorderLocalCostScore);
+            }
+        }
+    };
+
+    auto loopSPD = [&](const impl_desc_type type, int& selectedPrimitive, int& reorderCostScore) {
```
Comment (EgorDuplensky) [inline src/plugins/intel_cpu/src/node.cpp:373]: `selectSPDwithType`

Also let's avoid passing selectedPrimitiveIdx by reference.
Just return it by value.
I believe there is no reason to pass `reorderCostScore` as a parameter.
It can be just a local variable of the lambda.

File: src/plugins/intel_cpu/src/node.cpp [hunk 5]
```diff
@@ -316,6 +315,118 @@ void Node::selectPreferPrimitiveDescriptor(const std::vector<impl_desc_type>& pr
...
+                               " has more input ports than node: ",
+                               descInConfSize,
+                               " vs ",
+                               getParentEdges().size());
+                continue;
+            }
+
+            loopEdges(i, descInConfSize, supportedPrimitiveDesc, reorderLocalCostScore);
+
+            if (reorderLocalCostScore < reorderCostScore) {
```
Comment (EgorDuplensky) [inline src/plugins/intel_cpu/src/node.cpp:399]: Maybe something like `estimate < bestEstimate`

File: src/plugins/intel_cpu/src/node.cpp [hunk 6]
```diff
@@ -316,6 +315,118 @@ void Node::selectPreferPrimitiveDescriptor(const std::vector<impl_desc_type>& pr
     selectPrimitiveDescriptorByIndex(0);
 }
 
+bool Node::checkScalarShape(const ov::PartialShape& pshape) {
```
Comment (EgorDuplensky) [inline src/plugins/intel_cpu/src/node.cpp:318]: can we use an existing api instead?
Something like:
``` cpp
  return one_of(ov::shape_size(pshape), 0, 1);
```
There is chance we need to get Shape from a PartialShape first.

File: src/plugins/intel_cpu/src/node.cpp [hunk 7]
```diff
@@ -316,6 +315,118 @@ void Node::selectPreferPrimitiveDescriptor(const std::vector<impl_desc_type>& pr
...
+    for (auto s : pshape) {
+        if (s.is_static() && s.get_length() == 1) {
+            oneNum++;
+        }
+    }
+    return oneNum >= sz - 1;
+}
+
+void Node::selectPreferPrimitiveDescriptorWithShape(const std::vector<impl_desc_type>& priority, bool ignoreConstInputs) {
+    auto loopEdges = [&](size_t i,
```
[the remark this answers is not in the document]
Comment (xipingyan) [reply src/plugins/intel_cpu/src/node.cpp:330]: Updated, but still keep `i`, because I also want to keep the `DEBUG_LOG`.

File: src/plugins/intel_cpu/src/node.cpp (hunk 4 above)
Comment (xipingyan) [reply src/plugins/intel_cpu/src/node.cpp:373]: Updated.

File: src/plugins/intel_cpu/src/node.cpp (hunk 6 above)
Comment (xipingyan) [reply src/plugins/intel_cpu/src/node.cpp:318]: For example: {1,1,10} also is Scalar shape, `one_of(ov::shape_size(pshape), 0, 1)` seems can't meet our requirement.

Comment (EgorDuplensky) [reply src/plugins/intel_cpu/src/node.cpp]: `For example: {1,1,10} also is Scalar shape`
I believe it is not.
Currently the following shapes are threated as scalar ones:
- {}
- {1}
- {1, 1, 1, 1, ...}

Considering the fact, that you are looking for '{1,1,10}' shapes as well, now I am a bit confused what we are trying to achieve by checking for this type of shapes and estimating them as `1`. Could you please clarify?

Comment (xipingyan) [reply src/plugins/intel_cpu/src/node.cpp]: So confuse you.
Scalar shape self means only 1 element, but actually I want to find out 1 dim shape. I think the reorder of 1 dim shape also have less computation, because It will copy elements sequentially. What do you think? @EgorDuplensky 

1: Keep current logic, just rename scalar shape to 1 dim shape.
2: Exclude 1 dim shape, just check Scalar shape.(of course, it also can fix Zoom's regression issue)
image

2*(1*256*1*1) < (1*256*64*64)

Comment (EgorDuplensky) [reply src/plugins/intel_cpu/src/node.cpp]: Let's not use a term 'Scalar shape' for the tensor such as {1, 1, 1, 256}
This is misleading and confusing.

I just don't understand why we can't simply remove the special handling for such cases then?
I mean, in the example you have provided above, the generic logic which just computes `ov::shape_size()` would have also worked, isn't it?
Are there any other scenarios where the genetic logic actually would not work?

I can assume that such type of tensors ({1,1,X,1} or {1,1,1,X}) actually never require reorders, if the only difference in the descriptor is layout, so that precision is the same. That is why we might want to treat them differently. But if it is the case, then we better to implement some kind of check isReorderRequired(desc1, desc2), where we return false for such cases and an estimate for them would be 0 (not even 1)

Anyway, let's stick with the current implementation.
We can improve it later.

**Review by EgorDuplensky — APPROVED** [review]

**Comment by EgorDuplensky** [comment]
@dmitry-gorokhov Ready for further steps

File: src/plugins/intel_cpu/src/node.cpp (hunk 6 above)
Comment (xipingyan) [reply src/plugins/intel_cpu/src/node.cpp]: You are right, 
1: ov::shape_size() is good and generic logic, it also works. 
2: Add isReorderRequired
Updated.

File: src/plugins/intel_cpu/src/node.cpp [hunk 8]
```diff
@@ -316,6 +315,121 @@ void Node::selectPreferPrimitiveDescriptor(const std::vector<impl_desc_type>& pr
...
+                }
+                auto curDesc = supportedPrimitiveDesc.getConfig().inConfs[j].getMemDesc();
+                auto parentDesc = parent_spd->getConfig().outConfs[inNum].getMemDesc();
+
+                const bool isCompatible = curDesc->isCompatible(*parentDesc);
+                if (!isCompatible) {
+                    if (!isReorderRequired(parentDesc, curDesc)) {
+                        estimate += 1;
+                    } else {
+                        estimate += ov::shape_size<ov::intel_cpu::VectorDims>(curDesc->getShape().getStaticDims());
```
Comment (dmitry-gorokhov) [inline src/plugins/intel_cpu/src/node.cpp:364]: The algo assumes Subgraph works with static shapes only.
We need to add logic that clarifies the shape is not dynamic - overwise we will get exception here.

**Comment by dmitry-gorokhov** [comment]
@IvanNovoselov I would ask you yo take a look as well

File: src/plugins/intel_cpu/src/node.cpp (hunk 8 above)
Comment (xipingyan) [reply src/plugins/intel_cpu/src/node.cpp]: fix potential dynamic case exception issue, just call `getMinDims`

**Review by IvanNovoselov — APPROVED** [review]
No objections

**Comment by yuxu42** [comment]
Hi @EgorDuplensky  could you please take a look to see whether it can be merged? Thanks!

File: src/plugins/intel_cpu/src/node.cpp (hunk 8 above)
Comment (dmitry-gorokhov) [reply src/plugins/intel_cpu/src/node.cpp]: MinDims might contain UNDEFINED_DIM. Will the logic be correct in that case?

Comment (xipingyan) [reply src/plugins/intel_cpu/src/node.cpp]: I did a test, and found dynamic shape with UNDEFINED_DIM will return 0. 
```
auto s1 = ov::intel_cpu::Shape(ov::PartialShape({2, -1, 6}));
std::cout << "s1=" << s1.toString() << ",isDynamic=" << s1.isDynamic()
                          << ", getMinDims=" << ov::util::vector_to_string(s1.getMinDims())
                          << ", getMaxDims=" << ov::util::vector_to_string(s1.getMaxDims()) << std::endl;
```
Result:
`s1={2, 0 - ?, 6},isDynamic=1, getMinDims=[ 2, 0, 6 ], getMaxDims=[ 2, 18446744073709551615, 6 ]`

So I have to adopt old logical if current node has any dynamic input. @dmitry-gorokhov

File: src/plugins/intel_cpu/src/node.cpp [hunk 9]
```diff
@@ -334,6 +334,13 @@ bool Node::isReorderRequired(ov::intel_cpu::MemoryDescPtr desc1, ov::intel_cpu::
 }
 
 void Node::selectPreferPrimitiveDescriptorWithShape(const std::vector<impl_desc_type>& priority, bool ignoreConstInputs) {
+    // Filter out dynamic shape.
```
[the remark this answers is not in the document]
Comment (xipingyan) [reply src/plugins/intel_cpu/src/node.cpp:337]: Replaced with:
```
    if (isDynamic) {
        return selectPreferPrimitiveDescriptor(priority, ignoreConstInputs);
    }
```

**Review by dmitry-gorokhov — APPROVED** [review]
